### PR TITLE
[GEOT-6860] SchemaCache doesn't follow redirect when downloading from HttpConnection

### PR DESCRIPTION
**Pull request by taba90**
Jira ticket https://osgeo-org.atlassian.net/browse/GEOT-6860

this pr: 
- makes the SchemaCache able to follow a redirect when downloading;
- fix an https url in the online-test replacing a not working one causing the test to fail

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

File: modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java [hunk 1]
```diff
@@ -463,4 +465,17 @@ static boolean isSuitableDirectoryToContainCache(File directory) {
...
+    private static boolean hasRedirect(int responseCode) {
+        return responseCode == HttpURLConnection.HTTP_SEE_OTHER
+                || responseCode == HttpURLConnection.HTTP_MOVED_PERM
+                || responseCode == HttpURLConnection.HTTP_MOVED_TEMP;
+    }
+
+    private static byte[] followRedirect(HttpURLConnection connection, int blockSize)
+            throws URISyntaxException {
+        String redirect = connection.getHeaderField("Location");
+        URI redirectURL = new URI(redirect);
```
Comment (nmco) [inline modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java:478]: I would add a log message here at INFO level.

File: modules/library/xml/src/test/java/org/geotools/xml/resolver/SchemaCacheOnlineTest.java [hunk 2]
```diff
@@ -119,16 +121,16 @@ public void downloadHttps() throws Exception {
...
     private void check(byte[] bytes) {
-        Assert.assertTrue(bytes.length > 0);
+        assertTrue(bytes.length > 0);
         String text = new String(bytes);
-        Assert.assertTrue(text.contains("GeoSciML"));
-        Assert.assertTrue(text.contains("<schema"));
-        Assert.assertTrue(text.contains("</schema>"));
+        assertTrue(text.contains("GeoSciML"));
+        assertTrue(text.contains("<schema"));
+        assertTrue(text.contains("</schema>"));
```
Comment (nmco) [inline modules/library/xml/src/test/java/org/geotools/xml/resolver/SchemaCacheOnlineTest.java:128]: Looks like these are unrelated changes done by the IDE.

**Comment by nmco** [comment]
@taba90 should not the schema cache by delegating HTTP requests on GeoTools HTTP client? Then the HTTP client would be the one taking care of the redirect, fixing this issue for all the code that needs to performs an HTTP request, e.g. WMS client, WFS client, etc ...?

File: modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java (hunk 1 above)
Comment (taba90) [reply modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java]: fixed. Added an additional null check also

File: modules/library/xml/src/test/java/org/geotools/xml/resolver/SchemaCacheOnlineTest.java (hunk 2 above)
Comment (taba90) [reply modules/library/xml/src/test/java/org/geotools/xml/resolver/SchemaCacheOnlineTest.java]: fixed

File: modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java [hunk 3]
```diff
@@ -246,7 +246,9 @@ static void store(File file, byte[] bytes) {
             connection.setReadTimeout(downloadTimeout);
             connection.setUseCaches(false);
             connection.connect();
-            if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
+            int responseCode = connection.getResponseCode();
+            if (responseCode != HttpURLConnection.HTTP_OK) {
+                if (hasRedirect(responseCode)) return followRedirect(connection, blockSize);
```
Comment (aaime) [inline modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java:251]: While not very common, it's possible that the second response is another redirect. And so on.
Browsers follow them, but have some limits to avoid being caught in a redirect loop, e.g.:
https://www.liquidweb.com/kb/troubleshooting-too-many-redirects/

Comment (taba90) [reply modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java]: fixed. Seemed reasonable to me to put a max number of redirect of 5.

**Comment by taba90** [comment]
> @taba90 should not the schema cache by delegating HTTP requests on GeoTools HTTP client? Then the HTTP client would be the one taking care of the redirect, fixing this issue for all the code that needs to performs an HTTP request, e.g. WMS client, WFS client, etc ...?

thanks for your point @nmco. The SchemaCache seems to me the correct place to follow redirect, as long as it is it taking care of downloading the xsd, that looks to me more a "low level" operation that the HTTPClient of WMS WFS ecc, should no care about as it is meant more to be used OGC service specific request.

**Comment by taba90** [comment]
@nmco @aaime thanks for your reviews. Your feedbacks have been applied.

File: modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java [hunk 4]
```diff
@@ -463,4 +475,31 @@ static boolean isSuitableDirectoryToContainCache(File directory) {
...
+            try {
+                URI redirectURL = new URI(redirect);
+                LOGGER.info("Following redirect to " + redirect);
+                result = download(redirectURL, blockSize, redirectionCount);
+            } catch (URISyntaxException uri) {
+                LOGGER.warning(
+                        "Tried to follow redirect but invalid url was provided in Location header: "
+                                + redirect);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (taba90) [reply modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java:508]: fixed

File: modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java [hunk 5]
```diff
@@ -97,6 +97,9 @@
     /** Default download timeout. Change it with -Dschema.cache.download.timeout=<milliseconds> */
     private static int downloadTimeout = 60000;
 
+    /** Max number of attempts to follow a redirect to avoid redirect loop * */
+    private static final int MAX_FOLLOW_REDIRECT = 5;
```
[the remark this answers is not in the document]
Comment (taba90) [reply modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java]: fixed
